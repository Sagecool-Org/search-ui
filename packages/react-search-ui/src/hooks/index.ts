export { useSearch } from "./use_search";
