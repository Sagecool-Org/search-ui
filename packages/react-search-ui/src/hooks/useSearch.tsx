import { SearchContextState } from "@elastic/search-ui";
import { useContext, useEffect, useState } from "react";

import SearchContext from "../SearchContext";

/**
 * React hook that provides access to Search UI state and actions
 *
 * @param mapContextToProps Optional function to select specific parts of the context
 * @returns Selected search context state and actions
 */
export function useSearch<T = SearchContextState>(
  mapContextToProps?: (context: SearchContextState) => T
): T {
  const context = useContext(SearchContext);

  if (!context) {
    throw new Error("useSearch must be used within a SearchProvider");
  }

  const [state, setState] = useState<T>(() =>
    mapContextToProps
      ? mapContextToProps({
          ...context.driver.getState(),
          ...context.driver.getActions()
        })
      : ({
          ...context.driver.getState(),
          ...context.driver.getActions()
        } as T)
  );

  useEffect(() => {
    const subscription = (newState: Partial<SearchContextState>) => {
      setState((prevState: T) => {
        const fullContext = {
          ...(prevState as any),
          ...newState
        };
        return mapContextToProps
          ? mapContextToProps(fullContext)
          : (fullContext as T);
      });
    };

    context.driver.subscribeToStateChanges(subscription);
    return () => {
      context.driver.unsubscribeToStateChanges(subscription);
    };
  }, [context.driver, mapContextToProps]);

  return state;
}

export default useSearch;
