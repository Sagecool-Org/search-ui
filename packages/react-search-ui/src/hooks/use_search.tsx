import { createContext, useContext } from "react";

import type { SearchContextState } from "@elastic/search-ui";

export const UseSearchContext = createContext<SearchContextState | undefined>(
  undefined
);

export const useSearch = () => {
  const context = useContext(UseSearchContext);
  if (context === undefined) {
    throw new Error("useSearch must be used within a SearchContextProvider");
  }
  return context;
};

export default useSearch;
