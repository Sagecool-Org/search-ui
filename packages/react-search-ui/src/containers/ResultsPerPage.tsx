import React, { Component } from "react";
import { withSearch } from "..";
import {
  ResultsPerPage,
  ResultsPerPageContainerContext,
  ResultsPerPageContainerProps,
  ResultsPerPageViewProps
} from "@elastic/react-search-ui-views";
import { useSearch } from "../hooks";

export function ResultsPerPageContainer({
  className,
  options,
  view,
  ...rest
}: ResultsPerPageContainerProps) {
  const {
    driver: {
      state: { resultsPerPage },
      actions: { setResultsPerPage }
    }
  } = useSearch();
  const View = view || ResultsPerPage;
  const viewProps: ResultsPerPageViewProps = {
    className,
    onChange: (value) => {
      setResultsPerPage(value);
    },
    options,
    value: resultsPerPage,
    ...rest
  };

  return <View {...viewProps} />;
}

export default ResultsPerPageContainer;
