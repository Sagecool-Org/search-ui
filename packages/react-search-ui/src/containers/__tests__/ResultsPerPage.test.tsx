import React from "react";
import { shallow } from "enzyme";
import { ResultsPerPageContainer } from "../ResultsPerPage";
import type { ResultsPerPageViewProps } from "@elastic/react-search-ui-views";
import { useSearch } from "../../hooks";

jest.mock("../../hooks", () => ({
  useSearch: jest.fn()
}));

const params = {
  resultsPerPage: 20,
  setResultsPerPage: jest.fn()
};

beforeEach(() => {
  jest.clearAllMocks();

  (useSearch as jest.Mock).mockReturnValue(params);
});

it("supports a render prop", () => {
  const render = ({ value }: ResultsPerPageViewProps) => {
    return <div>{value}</div>;
  };
  const wrapper = shallow(<ResultsPerPageContainer view={render} />).dive();
  expect(wrapper).toMatchSnapshot();
});

it("will call back when a selection is made in the view", () => {
  let viewProps;

  shallow(
    <ResultsPerPageContainer
      {...params}
      view={(props) => {
        viewProps = props;
        return <div />;
      }}
    />
  ).dive();

  const { onChange } = viewProps;
  onChange(40);

  const resultsPerPage = params.setResultsPerPage.mock.calls[0][0];
  expect(resultsPerPage).toEqual(40);
});

it("passes className through to the view", () => {
  let viewProps;
  const className = "test-class";
  shallow(
    <ResultsPerPageContainer
      {...params}
      className={className}
      view={(props) => {
        viewProps = props;
        return <div />;
      }}
    />
  ).dive();
  expect(viewProps.className).toEqual(className);
});

it("renders the component with custom page options", () => {
  const options = [5, 10, 15];
  const resultsPerPage = 10;
  (useSearch as jest.Mock).mockReturnValue({
    ...params,
    resultsPerPage
  });
  const wrapper = shallow(<ResultsPerPageContainer options={options} />).dive();

  expect(wrapper).toMatchSnapshot();
});

it("passes data-foo through to the view", () => {
  let viewProps;
  const data = "bar";
  shallow(
    <ResultsPerPageContainer
      data-foo={data}
      view={(props) => {
        viewProps = props;
        return <div />;
      }}
    />
  ).dive();
  expect(viewProps["data-foo"]).toEqual(data);
});
