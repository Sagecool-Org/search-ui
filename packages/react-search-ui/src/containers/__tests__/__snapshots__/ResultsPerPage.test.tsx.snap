// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component with custom page options 1`] = `
<div
  className="sui-results-per-page"
>
  <div
    className="sui-results-per-page__label"
  >
    Show
  </div>
  <ForwardRef
    className="sui-select sui-select--inline"
    classNamePrefix="sui-select"
    components={
      Object {
        "Option": [Function],
      }
    }
    isSearchable={false}
    onChange={[Function]}
    options={
      Array [
        Object {
          "label": 5,
          "value": 5,
        },
        Object {
          "label": 10,
          "value": 10,
        },
        Object {
          "label": 15,
          "value": 15,
        },
      ]
    }
    styles={
      Object {
        "control": [Function],
        "dropdownIndicator": [Function],
        "indicatorSeparator": [Function],
        "option": [Function],
        "singleValue": [Function],
        "valueContainer": [Function],
      }
    }
    value={
      Object {
        "label": 10,
        "value": 10,
      }
    }
  />
</div>
`;

exports[`supports a render prop 1`] = `
<div>
  20
</div>
`;
